removing useless variables && declaring missed ones

Jabher opened the pull request:
Guys, I've detected some minor issues:
- some variables are falling into global scope (e.g. `for (level in layers) {` - `var level` should be here instead, `var iterations = bucketSize = 0;` - `bucketSize` should be declared separately);
- some variables are not used (e.g. `var type = connection.gatedfrom[from].type;`, `var xtrace = this.trace.extended[id];`)
- some code style issues - semicolons and so on.

Also there are significant possible issues regarding `for... in` loops - libraries such as prototype.js, sugar.js and several others extends the prototype and will possibly cause an iteration over an non-own properties. There are several ways to solve this, so if you'd like to, I can work over this question.

cazala commented:
Looks like the optimized networks don't behave the same way as the unoptimized ones after these changes, I ll take a look and and try to see why.

Jabher commented:
Sounds a bit strange. Can you publish tests you've run? I can try to figure
out what's broken.

пн, 18 Июл 2016, 16:41 Juan Cazala <notifications@github.com>:

> Looks like the optimized networks don't behave the same way as the
> unoptimized ones after these changes, I ll take a look and and try to see
> why.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cazala/synaptic/pull/115#issuecomment-233331403>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABNfCPaj4cYipHY3y9xpqznvo01BCVz9ks5qW4KGgaJpZM4JOVFT>
> .
>
-- 

---
Vsevolod Rodionov

cazala commented:
You can find the source of the tests [here](https://github.com/cazala/synaptic/blob/master/test/synaptic.js) and the execution of them [here]
(https://travis-ci.org/cazala/synaptic/builds/145420053). To run them locally just use `npm run test`.

cazala commented:
This only test failing is [this one](https://github.com/cazala/synaptic/blob/master/test/synaptic.js#L432-L462) tho

Jabher commented:
That's... wow. First I thought it's happening because I'm using latest node.js. I've switched to 0.10 like in travis.yml. It did not help.
Then I've decided that it's happening because I'm using mac. I've switched to docker's alpine-node. Everything was fine both in latest and 0.10.
Then I've decided to use travis. To cover more edge cases and trigger Travis I've added more travis envs. And IT PASSED EVERYTHING.
Both here and in my local travis - https://travis-ci.org/Jabher/synaptic/builds/145816086 - everything is fine.

I'm really confused.

cazala commented:
It looks like it's passing now. The only missing thing are the dist files, could you run `npm run build` and commit that?

Jabher commented:
done :) or you want squashed commit?

Jabher commented:
wow, it reproduced again. Give me some time to figure out what's happening.

Jabher commented:
So, I was able to reproduce the error. 
It is reproducible only in node 0.10 and is bringing a `mse` value a little higher than expected.
In my case it was 1.0794810088851979e-10 (with threshold of 1e-10).
I'll rewrite the test to catch this case better. Also threshold can be raised or I can investigate why this is happening.

Jabher commented:
So, I think I figured out what's happened.

First of all, tests were broken. I fixed them - networks comparison was always running against 1000 iterations (even on 200, 400 and so on iterations were used). For small amount of iterations vector distance (MSE) is much more significant rather than 1e-10.
Don't be afraid that number of tests became smaller - I've removed test runs for low iterations count, as they were running pointlessly.

Secondly - old node.js was using not so crypto-secure Math.random() (http://v8project.blogspot.ru/2015/12/theres-mathrandom-and-then-theres.html), so I have a feeling that random function changed everything dramatically. For node 0.10 I was receiving mean square difference for opt/deopt functions between 1e-13 and 3e-9, on all next versions of node.js it's always below 2e-11. I'm not quite sure how to handle this, but probably I'd rather just drop node 0.10 and 0.12 support as they are in maintenance now (https://github.com/nodejs/LTS) and will be dropped soon - in october and december respectfully. By dropping I mean just removing them from test scope in travis.yml.


I can separate everything into separate merge requests, however, you probably should review the tests anyway.

cazala commented:
Looking great (: do you wanna remove 0.10 and 0.12 from `.travis.yml` before merging?

Jabher commented:
My pleasure!

just a small note: I've first tested it with all available node.js 4+ versions (around 20 instances), it was very slowly passing all tests, so I've left only 4, 5, 6 and latest versions in travis.yml.

Jabher commented:
Aaaand I have a feeling I've broken Travis CI  😞

cazala commented:
It looks good to me, what do you mean it's broken?

Jabher commented:
It was hanging for half an hour or so :) And it was feeling like it was totally not responsible. 

Commit itself was fine, I was just a little bit afraid I've made Travis watchdog hang up.

cazala commented:
Ahh I see :P it's looking great to me, I'm merging. Thank you !

Jabher commented:
great! 🍻

I'll make a little more PRs soon, I think.

cazala commented:
keep em' coming! :D
